torch/distributed/fsdp/sharded_grad_scaler.py

```diff
@@ -118,45 +98,29 @@ def scale(self, outputs: Iterable[torch.Tensor]) -> Iterable[torch.Tensor]:
-            scaled_output = outputs * self._scale.to(
-                device=outputs.device, non_blocking=True
+        scaled_outputs = super().scale(outputs)
+        # Here we ensure the return dtype is the same as the outputs dtype.
+        # For the FSDP + Mixed Precision use case, the loss output is in the Mixed Precision
+        # format (fp16, bf16) and so the scaled loss should be of the same dtype.
+        if isinstance(scaled_outputs, torch.Tensor):
+            return scaled_outputs.type(outputs.dtype)  # type: ignore[union-attr]
+
+        def apply_dtype(
+            scaled_val: Union[torch.Tensor, Iterable[torch.Tensor]],
+            orig_val: Union[torch.Tensor, Iterable[torch.Tensor]],
+        ):
+            if isinstance(scaled_val, torch.Tensor):
+                return scaled_val.type(orig_val.dtype)  # type: ignore[union-attr]
+            if isinstance(scaled_val, abc.Iterable):
+                iterable = map(apply_dtype, scaled_val, orig_val)
+                if isinstance(scaled_val, (list, tuple)):
+                    return type(scaled_val)(iterable)
+                return iterable
+            raise ValueError(
+                "scaled_outputs/outputs must be a Tensor or an iterable of Tensors"
             )
-            # Here we ensure the return dtype is the same as the outputs dtype.
-            # For the FSDP + Mixed Precision use case, the loss output is in the Mixed Precision
-            # format (fp16, bf16) and so the scaled loss should be of the same dtype.
-            return scaled_output.type(outputs.dtype)
-
-        stash: List[_GeneralMultiDeviceReplicator] = []
-
-        def apply_scale(val: Union[torch.Tensor, Iterable[torch.Tensor]]):
-            if isinstance(val, torch.Tensor):
-                assert _is_supported_device(val)
-                if len(stash) == 0:
-                    if self._scale is None:
-                        self._lazy_init_scale_growth_tracker(val.device)
-                    assert self._scale is not None
-                    stash.append(_GeneralMultiDeviceReplicator(self._scale))
-                scaled_val = val * stash[0].get(val.device)
-                # Here we ensure the return dtype is the same as the outputs dtype.
-                # For the FSDP + Mixed Precision use case, the loss output is in the Mixed Precision
-                # format (fp16, bf16) and so the scaled loss should be of the same dtype.
-                return scaled_val.type(val.dtype)
-            if isinstance(val, abc.Iterable):
-                iterator = map(apply_scale, val)
-                if isinstance(val, (list, tuple)):
-                    return type(val)(iterator)
-                return iterator
-            raise ValueError("outputs must be a Tensor or an iterable of Tensors")
-
-        return apply_scale(outputs)
+
+        return apply_dtype(scaled_outputs, outputs)
 
     def _foreach_non_finite_check_and_unscale_cpu_(
```

Comment:
I would rename this to not include "foreach" in the name as this doesn't actually do a horizontal fusion on CPU